src/tools/rbd_mirror/image_map/Policy.cc

```diff
@@ -0,0 +1,200 @@
+    images->emplace(image_id.id, image_id.global_id, instance_id);
+  }
+
+  if (mapped.empty()) {
+    on_finish->complete(0);
+    return;
+  }
+
+  UpdateRequest<I> *req = UpdateRequest<I>::create(
+    m_ioctx, mapped, on_finish);
+  req->send();
+}
+
+template <typename I>
+void Policy<I>::update_images_removed(const std::string &mirror_uuid,
+                                      ImageIds &&image_ids, Images *images,
+                                      Context *on_finish) {
+  // images that need to be purged -- once we start maintaining
+  // state for images, image map remove request can selectively
+  // purge on-disk map.
+  GlobalImageIds purge;
+
+  for (auto const &image_id : image_ids) {
+    dout(20) << ": image_id=" << image_id << dendl;
+
+    std::string instance_id = lookup(image_id.global_id);
+    assert(instance_id != UNMAPPED_INSTANCE_ID);
+
+    images->emplace(image_id.id, image_id.global_id, instance_id);
+
+    auto rm = m_info[image_id.global_id].peer_uuid.erase(mirror_uuid);
+    assert(rm == 1);
+
+    if (m_info[image_id.global_id].peer_uuid.empty()) {
+      unmap(image_id.global_id);
+      m_info.erase(image_id.global_id);
+
+      purge.emplace(image_id.global_id);
+    }
+  }
+
+  if (purge.empty()) {
+    on_finish->complete(0);
+    return;
+  }
+
+  RemoveRequest<I> *req = RemoveRequest<I>::create(
+    m_ioctx, purge, on_finish);
+  req->send();
+}
+
+template <typename I>
+void Policy<I>::init(Context *on_finish) {
+  dout(20) << dendl;
+
+  LoadRequest<I> *req = LoadRequest<I>::create(
+    m_ioctx, &m_inmap, on_finish);
+  req->send();
+}
+
+template <typename I>
+void Policy<I>::shut_down(Context *on_finish) {
+  dout(20) << dendl;
+
+  m_async_op_tracker.wait_for_ops(on_finish);
+}
+
+template <typename I>
+void Policy<I>::update_images(const std::string &mirror_uuid,
+                              ImageIds &&added_image_ids,
+                              ImageIds &&removed_image_ids,
+                              MapSpecRef mapspec, Context *on_finish) {
+  dout(20) << ": mirror_uuid=" << mirror_uuid << ", "
+           << "added_count=" << added_image_ids.size() << ", "
+           << "removed_count=" << removed_image_ids.size() << dendl;
+
+  m_async_op_tracker.start_op();
+  on_finish = new FunctionContext([this, on_finish](int r) {
+      dout(20) << ": r=" << r << dendl;
+      m_async_op_tracker.finish_op();
+      on_finish->complete(r);
+    });
+
+  C_Gather *gather_ctx = new C_Gather(g_ceph_context, on_finish);
+
+  {
+    RWLock::WLocker locker(m_lock);
+
+    if (!removed_image_ids.empty()) {
+      update_images_removed(mirror_uuid, std::move(removed_image_ids),
+                            &mapspec->removed, gather_ctx->new_sub());
+    }
+    if (!added_image_ids.empty()) {
+      update_images_added(mirror_uuid, std::move(added_image_ids),
+                          &mapspec->added, gather_ctx->new_sub());
+    }
+  }
```

Comment:
This was intended to be called via `PoolWatcher` so it was safe I guess since upon refreshing local and remote pool we would know what images in the local pool needs to be removed. This would not be true if its invoked via `PoolReplayer`.